std/range/primitives.d

```diff
@@ -1987,6 +1993,177 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
 }
 
 /**
+Pops the first element from $(D r) and returns it.
+It combines $(D front) and $(D popFront) in one method.
+
+Params:
+    r = Forward range
+
+Returns:
+    Next element of the input range
+ */
+auto frontPop(Range)(ref Range r)
+if (isForwardRange!Range)
+in
+{
+    assert(!r.empty, "empty range passed to " ~ __PRETTY_FUNCTION__);
+}
+body
+{
+    auto e = r.save.front;
```

Comment:
What we would need is a version that handles forward ranges that return reference-containing types, and one that handles input ranges that return value types.